src/picom.c

```diff
@@ -2174,6 +2273,7 @@ static void session_destroy(session_t *ps) {
 	c2_list_free(&ps->o.paint_blacklist);
 	c2_list_free(&ps->o.unredir_if_possible_blacklist);
 	c2_list_free(&ps->o.rounded_corners_blacklist);
+	c2_list_free(&ps->o.window_shader_fg_rules);
```

Comment:
We leak the filename allocated by `realpath()` in the `load_auxiliary_file*()` functions, that is stored as the userdata in the rule.

I suggest the following patch:
```diff
diff --git a/src/c2.c b/src/c2.c
index e557f29..437544f 100644
--- a/src/c2.c
+++ b/src/c2.c
@@ -1693,3 +1693,10 @@ bool c2_list_foreach(const c2_lptr_t *condlist, c2_list_foreach_cb_t cb, void *d
 void *c2_list_get_data(const c2_lptr_t *condlist) {
 	return condlist->data;
 }
+
+/// Free user data stored in a condition.
+void c2_list_free_data(const c2_lptr_t *condlist) {
+	if (condlist->data) {
+		free(condlist->data);
+	}
+}
diff --git a/src/c2.h b/src/c2.h
index 1f3390b..1c0a273 100644
--- a/src/c2.h
+++ b/src/c2.h
@@ -30,6 +30,8 @@ typedef bool (*c2_list_foreach_cb_t)(const c2_lptr_t *cond, void *data);
 bool c2_list_foreach(const c2_lptr_t *list, c2_list_foreach_cb_t cb, void *data);
 /// Return user data stored in a condition.
 void *c2_list_get_data(const c2_lptr_t *condlist);
+/// Free user data stored in a condition.
+void c2_list_free_data(const c2_lptr_t *condlist);
 
 /**
  * Destroy a condition list.
diff --git a/src/picom.c b/src/picom.c
index 047f5e7..7bed6b1 100644
--- a/src/picom.c
+++ b/src/picom.c
@@ -1604,6 +1604,11 @@ static bool load_shader_source_for_condition(const c2_lptr_t *cond, void *data)
 	return load_shader_source(data, c2_list_get_data(cond));
 }
 
+static bool free_shader_source_for_condition(const c2_lptr_t *cond, void *data attr_unused) {
+	c2_list_free_data(cond);
+	return false;
+}
+
 /**
  * Initialize a session.
  *
@@ -2273,6 +2278,9 @@ static void session_destroy(session_t *ps) {
 	c2_list_free(&ps->o.paint_blacklist);
 	c2_list_free(&ps->o.unredir_if_possible_blacklist);
 	c2_list_free(&ps->o.rounded_corners_blacklist);
+
+	// Free window shader rules
+	c2_list_foreach(ps->o.window_shader_fg_rules, free_shader_source_for_condition, NULL);
 	c2_list_free(&ps->o.window_shader_fg_rules);
 
 	// Free tracked atom list
 ```
Alternatively, we can change the opacity-rule to use an allocated `long` for its parameter and always free the userdata in `c2_list_free()`. We currently cannot blindly `free()` the userdata because the opacity-rules use the `void *` as storage for a `long`.

Reply:
I thought `c2_list_free` would free the user data :(. Turns out the user data pointer never stored allocated memory until now.